Introduce IoHandleEventLoopGroup / IoHandleEventLoop and its implemen…

normanmaurer opened the pull request:
…tations MultiThreadIoHandleEventLoopGroup / SingleThreadIoHandleEventLoop

Motivation:

At the moment each transport has its own EventLoopGroup and EventLoop implementation. Most of these are final because they are tightly coubled with the specific Channel implementations. This is unfortunate for multiple reasons, as for example:

- It's hard to instrument EventLoop* implementations as these are not extensible at all or even if they are the same logic needs to be added to multiple EventLoop* implementations.
- Registration / Deregistration is limited to Channels which remove flexibility as for example some Operation Systems allowing to also handle other IO (for example file IO) with something like an EventLoop (one example would be io_uring).
- A lot of code-duplication between different EventLoop* implementations exists as most share kind of the same logic of running non IO tasks.

Modifications:

- Add the concept of IoHandleEventLoopGroup / IoHandleEventLoop and the corrosponding IoHandle, IoHandler and IoHandlerFactory. This allows to only have a custom implement for IO handling per transport while share the same code for execution of tasks, scheduled task and ultimatly IO.
- Add the implementations MultiThreadIoHandleEventLoopGroup and SingleThreadIoHandleEventLoop that can be used by all (or most) transport implementations by having these just provide their own IoHandler / IoHandlerFactory implementations.

Result:

More code sharing and easier of customizing EventLoops.

normanmaurer commented:
Netty backport


364a5f843720b2e1fce60df9ca30f2f09159b419
8fd955a7e981ec141ab7d51dfaaf4b7d51eda1dd
47a18b193428548e4c2f031beac21af3400aae90


java8+
ee54ea725ad0b3afa7c7bea35cc1bf77a63234de
cba3b4dd57824039bae82ec064fdbf276c26d7cf


KQueueEventLoopGroup args ordering


I am opening this a draft mainly because I wanted to show and proof that we can back port some of the changes that we did in main without breaking too much things. I think even thought this PR introduce some API breaking it shouldn’t affect almost any user. To proof this I did not change how we create the different EventLoopGroups in our testsuite and everything still works. 

So while the old way of constructing Eventloops still work the new way would be a bit different:

Old API:
```
new EpollEventLoop()
```

New API:
```
new MultiThreadIoHandleEventLoopGroup(EpollHandler.newFactory();
````

The benefit of the new way is that it is quite easy to instrument / doctorate an EventLoop/EventLoopGroup now. Basically you can just extend MultiThreadIoHandleEventLoopGroup / SingleThreadIoHandleEventLoop and add all the customization once while be able to reuse it with any transport implementation.
This includes things like:
 - able to get a count of registered channels / handles
 - understand how long the io processing vs task processing took.
 - how many channels / handles were processed per loop run
 - customize / decorate promises
 - …

Beside this IoHandleEventLoop can support not only Channel but also other subtypes of IoHandle. This would for example allow us to add file io support with io_uring and still use the same eventloop as for sockets etc. I am sure there are more things we could do by adding this kind of flexibility.

With all this said I still think the change is too dangerous for a new 4.1.x release. With this in mind I wondered if we might want to aim for releasing 4.2.0 with the following change:

- This PR
- Change the minimum java version to 8
- Maybe a few more low risk changes.

I would love to hear what other project members and the community think about this ?

normanmaurer commented:
@bryce-anderson @idelpivnitskiy @chrisvest @franz1981 @vietj @yawkat @trustin WDYT ?

normanmaurer commented:
If interested I can also change all the creation code of `EventLoopGroup`s in the testsuite and examples. I just wanted to not do this for now to show how small the impact for most people should be

He-Pin commented:
Do you plan to align with the design in Netty 5?

yawkat commented on transport-classes-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:
```diff
@@ -16,142 +16,46 @@
-    private static final InternalLogger logger = InternalLoggerFactory.getInstance(EpollEventLoop.class);
-    private static final long EPOLL_WAIT_MILLIS_THRESHOLD =
-            SystemPropertyUtil.getLong("io.netty.channel.epoll.epollWaitThreshold", 10);
+    private int ioRatio;
```
This is now unused?

yawkat commented on transport-classes-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoopGroup.java:
```diff
@@ -157,37 +162,32 @@ public EpollEventLoopGroup(int nThreads, Executor executor, EventExecutorChooser
-        for (EventExecutor e: this) {
-            ((EpollEventLoop) e).setIoRatio(ioRatio);
-        }
+        // NOOP
```
Same here, need to document that this does nothing now

yawkat commented on transport-classes-epoll/src/main/java/io/netty/channel/epoll/EpollHandler.java:
```diff
@@ -0,0 +1,584 @@
+    /**
+     * Returns the percentage of the desired amount of time spent for I/O in the event loop.
+     */
+    public int getIoRatio() {
```
This new property is also unused

yawkat commented on transport/src/main/java/io/netty/channel/IoExecutionContext.java:
```diff
@@ -0,0 +1,39 @@
+
+/**
+ * The execution context for an {@link IoHandler}.
+ * All method must be called from the {@link EventLoop} thread.
```
```suggestion
 * All methods must be called from the {@link EventLoop} thread.
```

He-Pin replied:
The same typo exits on main too

normanmaurer replied:
I will "forward port" all changes that make sense to main

yawkat commented on transport/src/main/java/io/netty/channel/IoHandler.java:
```diff
@@ -0,0 +1,75 @@
+     * @param handleType the type of the {@link IoHandle}.
+     * @return if compatible of not.
+     */
+    boolean isCompatible(Class<? extends IoHandle> handleType);
```
i wonder if using reflection here will cause native-image issues

He-Pin replied:
at least we not cache it. eg: https://github.com/eclipse-vertx/vert.x/pull/5169

yawkat commented on transport/src/main/java/io/netty/channel/nio/NioHandler.java:
```diff
@@ -0,0 +1,688 @@
+    // - https://bugs.openjdk.java.net/browse/JDK-6527572 for JDK prior to 5.0u15-rev and 6u10
+    // - https://github.com/netty/netty/issues/203
+    static {
+        if (PlatformDependent.javaVersion() < 7) {
```
If we move to 8, we can remove workarounds like this

normanmaurer commented:
Will look into the build timeout later today

He-Pin commented:
> > Do you plan to align with the design in Netty 5?
> 
> What design you talk about ? The ideas was to introduce the same concept with minimal real world breakage 

I mean split the ioHandler with the event loop, I will take some time to look into this.

normanmaurer commented:
> > > Do you plan to align with the design in Netty 5?
> > 
> > 
> > What design you talk about ? The ideas was to introduce the same concept with minimal real world breakage
> 
> I mean split the ioHandler with the event loop, I will take some time to look into this.

That's exactly what this PR does ;)

normanmaurer commented:
Also I think if we would switch to java8 I could maybe do something more flexible to support different `IoHandle` in the future.

violetagg commented:
@normanmaurer I don't know whether we should discuss this here in this PR but


> With all this said I still think the change is too dangerous for a new 4.1.x release. With this in mind I wondered if we might 
> want to aim for releasing 4.2.0 with the following change:

> This PR
> Change the minimum java version to 8
> Maybe a few more low risk changes.


Have you considered moving out of incubation some the projects? (io_uring, QUIC, HTTP/3)?

normanmaurer commented on transport/src/main/java/io/netty/channel/IoHandleEventLoop.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * {@link EventLoop} that allows to register / deregister {@link IoHandle} instances.
+ */
+public interface IoHandleEventLoop extends EventLoop, IoHandleEventLoopGroup {
```
I wonder if the name is too long and we should just call it `IoEventLoop` and so follow the same pattern for others like:

- `IoEventLoopGroup`
- `SingleThreadIoEventLoopGroup`
- `MultiThreadIoEventLoopGroup`

This would make instancing these a bit less verbose

chrisvest replied:
Makes sense, so the `Io` part is the common theme between `IoHandle` and `IoEventLoop`. I'm fine either way, though.

normanmaurer replied:
Exactly... let's do it

normanmaurer commented:
> @normanmaurer I don't know whether we should discuss this here in this PR but
> 
> > With all this said I still think the change is too dangerous for a new 4.1.x release. With this in mind I wondered if we might
> > want to aim for releasing 4.2.0 with the following change:
> 
> > This PR
> > Change the minimum java version to 8
> > Maybe a few more low risk changes.
> 
> Have you considered moving out of incubation some of the projects? (io_uring, QUIC, HTTP/3)?

Yes but let's focus on this one first :)

He-Pin commented:
I would like to +1 for moving to Java 8 on Netty 4.2.x, as Netty 4.1.x has a very long history, With Java 8's default method, lambda completableFuture, etc., we can achieve more than it is now.

He-Pin commented on pom.xml:
```diff
@@ -1345,6 +1345,211 @@
+                  <code>java.class.nonPublicPartOfAPI</code>
+                  <new>interface io.netty.channel.epoll.EpollRegistration</new>
+                </item>
+
```
this is huge, epic.

normanmaurer replied:
While the list to ignore is very big I think in real world it will not affect people

normanmaurer commented on transport/src/main/java/io/netty/channel/nio/NioHandler.java:
```diff
@@ -0,0 +1,688 @@
+     * the select method and the select method will block for that time unless
+     * waken up.
+     */
+    private final AtomicBoolean wakenUp = new AtomicBoolean();
```
I think it not matters in terms of memory usage... so yep

He-Pin commented on transport-classes-epoll/src/main/java/io/netty/channel/epoll/EpollRegistration.java:
```diff
@@ -0,0 +1,46 @@
+    /**
+     * Returns an {@link IovArray} that can be used for {@code writev}.
+     */
+    IovArray cleanIovArray();
```
If we move to Java 8, this IovArray will extend a predicate

He-Pin commented on transport/src/main/java/io/netty/channel/nio/NioSelectableChannelHandle.java:
```diff
@@ -0,0 +1,103 @@
+        return nioProcessor;
+    }
+
+    protected abstract void handle(S channel, SelectionKey key);
```
By move to java 8 we have less diff with the main branch.

He-Pin replied:
This method can be abstracted with a lambda now as Netty 5

normanmaurer commented:
> I checked some of the files not all. I suggest first moving to Java 8 and then back to this PR.
> 
> After we moved to Java 8, the diff set of this PR against the main branch will be smaller.

I am not sure about that. But yeah we could start with a new PR moving to 8.

He-Pin commented:
Almost all services are running on Java 8, 11% on Java 11, and 0.05% on Java 21 here, not know about other companies, so this is a good time/reason for you to cut a Java 8 release, will reduce maintenance burden too.

He-Pin commented on transport-classes-epoll/src/main/java/io/netty/channel/epoll/EpollHandler.java:
```diff
@@ -0,0 +1,566 @@
+                            pendingWakeup = true;
+                        }
+                    }
+                    // fallthrough
```
fall-through

He-Pin commented on transport/src/main/java/io/netty/channel/nio/NioHandler.java:
```diff
@@ -0,0 +1,688 @@
+
+    int selectNow() throws IOException {
+        return selector.selectNow();
+    }
```
It's something like 
```java
    private int selectNow() throws IOException {
        try {
            return selector.selectNow();
        } finally {
            // restore wakeup state if needed
            if (wakenUp.get()) {
                selector.wakeup();
            }
        }
    }
```
on main

chrisvest commented on transport-classes-kqueue/src/main/java/io/netty/channel/kqueue/KQueueEventLoopGroup.java:
```diff
@@ -155,37 +165,70 @@ public KQueueEventLoopGroup(int nThreads, Executor executor, EventExecutorChoose
-        for (EventExecutor e: this) {
-            ((KQueueEventLoop) e).setIoRatio(ioRatio);
-        }
+        // noop
```
Also missing a documentation update here.

normanmaurer commented:
There are some more changes coming... stay tuned.

He-Pin commented:
@normanmaurer let's keep those design in sync with main:) but since the `io.netty5` package, then not easy for cherry-pick

normanmaurer commented:
I will port things over for for sure

normanmaurer commented on transport-classes-epoll/src/main/java/io/netty/channel/epoll/EpollHandler.java:
```diff
@@ -0,0 +1,566 @@
+/**
+ * {@link EventLoop} which uses epoll under the covers. Only works on Linux!
+ */
+public class EpollHandler implements IoHandler {
```
Yes it needs to be... that said we should make it final. All the `IoHandler` implementations need to be public as you use these to create the factory to inject in the `MultiThreadIoEventLoopGroup`.

For example:

```
var group = new MultiThreadIoEventLoopGroup(NioHandler.newFactory());
```

Another possibility would be to not make these public but provide a `IoHandlerFactory` implementation that is public.

Like this:
```
var group = new MultiThreadIoEventLoopGroup(new NioHandlerFactory());

```

This factory would then instance the handles. 

@chrisvest @He-Pin @yawkat not sure what is better ?

yawkat replied:
current approach seems fine to me

bryce-anderson replied:
I see. The factory seems nicer as it limits the escape of additional API but it's not a huge deal either way.

He-Pin replied:
As we are using Java 8 now, so final class with factory would be better.

normanmaurer commented on transport/src/main/java/io/netty/channel/nio/NioHandler.java:
```diff
@@ -0,0 +1,695 @@
+ * {@link Selector}.
+ *
+ */
+public final class NioHandler implements IoHandler {
```
see above

normanmaurer commented on transport/src/main/java/io/netty/channel/local/LocalIoHandler.java:
```diff
@@ -0,0 +1,108 @@
+import java.util.Set;
+import java.util.concurrent.locks.LockSupport;
+
+public final class LocalIoHandler implements IoHandler {
```
see above

normanmaurer commented on transport-classes-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:
```diff
@@ -260,328 +65,26 @@ private static Queue<Runnable> newTaskQueue0(int maxPendingTasks) {
-        } catch (IOException e) {
-            logger.warn("Failed to close the epoll fd.", e);
-        }
+    public void setIoRatio(int ioRatio) {
     }
```
yes that is a good idea

normanmaurer commented:
@He-Pin  @violetagg @bryce-anderson @idelpivnitskiy @chrisvest @franz1981 @vietj @yawkat @trustin I just pushed more changes... I think from an API point of view I am now consider this done (and everything compiles / pass tests).

So what changed from the original PR design:
 - Rename *IoHandleEventLoop* to *IoEventLoop* to make things less verbose
 - renamed method `registerForIo(...)` to just `register(...)` and let it return an `IoRegistration`. This registration can then be used to modify which events some is interested in.
 - `Channel` does not extend `IoHandle` anymore.
 - Added IoOpt interface and added implements for all transports. These `IoOpt` implementations can be used with the `IoRegistration` to manipulate for which events an `IoHandle` is interested to get notified.
 - `IoHandle` has now a `handle(....)` method that is called for the events and so we could decouple the specific `IoHandler` implementations completely from the respective `Channel` implementations.
 -  Added deprecations

The nice thing about the new API is really that it is trivial for advanced users to write their own `IoHandle` implementations and register it with our existing implementations without the need to touch either the `*IoEventLoop*`s or the `IoHandler` implementations for the various transports. This allows a lot of more advanced use cases and would for example allow users to write their own `File` support for `io_uring` once we use the new API there without the need for us to do it ourselves and a lot of other things.

The concept that is used now with `IoHandler` and `IoHandle` is kind of similar to what you get when using `libev`. So it's a lower-level API that can be used and is also used by ourselves.

I know the changes are non-trivial so thanks to everyone who takes the time for review and providing feedback :)

chrisvest commented on transport/src/main/java/io/netty/channel/nio/AbstractNioChannel.java:
```diff
@@ -77,9 +84,13 @@ public void run() {
     protected AbstractNioChannel(Channel parent, SelectableChannel ch, int readInterestOp) {
+        this(parent, ch, NioIoOpt.valueOf(readInterestOp));
+    }
+    protected AbstractNioChannel(Channel parent, SelectableChannel ch, NioIoOpt readOpt) {
```
Missing line break

chrisvest commented on transport/src/main/java/io/netty/channel/nio/NioIoHandler.java:
```diff
@@ -0,0 +1,753 @@
+/*
+ * Copyright 2012 The Netty Project
```
Update copyright year

chrisvest reviewed (commented):
A few comments

normanmaurer commented on transport/src/main/java/io/netty/channel/IoHandler.java:
```diff
@@ -0,0 +1,70 @@
+     * Wakeup the {@link IoHandler}, which means if any operation blocks it should be unblocked and
+     * return as soon as possible.
+     */
+    void wakeup(boolean inEventLoop);
```
Good idea... let us pass in `IoEventLoop`

normanmaurer commented:
Thanks a lot for all the reviews.... This is now merged :)
